File: src/783. Minimum Distance Between BST Nodes/main.py
# Solved by @kitanoyoru

import sys

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self) -> None:
        self.ans: int = sys.maxsize 
        self.prev: Optional[int] = None
        
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.ans

    def dfs(self, root: Optional[TreeNode]) -> None:
        if root is None:
            return
        self.dfs(root.left)
        if self.prev is not None:
            self.ans = min(self.ans, root.val - self.prev)
        self.prev = root.val
        self.dfs(root.right)

File: src/783. Minimum Distance Between BST Nodes/main.go
// Solved by @kitanoyoru

package main

import "math"

type TreeNode struct {
  Val int
  Left *TreeNode
  Right *TreeNode
}

func min(x, y int) int {
  if x > y {
    return y
  }
  return x
}

func dfs(node *TreeNode, ans, prev *int) {
  if node == nil {
    return
  }
  dfs(node.Left, ans, prev)
  if *prev != -1 {
    *ans = min(*ans, node.Val - *prev)
  }
  *prev = node.Val
  dfs(node.Right, ans, prev)
}

func minDiffInBST(root *TreeNode) int {
  ans, prev := math.MaxInt, -1 
  dfs(root, &ans, &prev)
  return ans
}
